frontend/lib/src/WidgetStateManager.ts

```diff
@@ -709,6 +719,47 @@ export class WidgetStateManager {
       this.props.formsDataChanged(this.formsData)
     }
   }
+
+  /**
+   * Get the element state value for the given element ID and key, if it exists.
+   * This is a frontend-only state that is never sent to the server.
+   */
+  public getElementState(elementId: string, key: string): any {
+    return this.elementStates.get(elementId)?.get(key)
+  }
+
+  /**
+   * Sets the state of an element identified by its ID and its key.
+   * This is a frontend-only state that is never sent to the server.
+   * It can be used to store element state to restore the state
+   * of an element in situations where an element is removed and re-added.
+   *
+   * @param {string} elementId - The unique identifier of the element.
+   * @param {string} key - The key to set
+   * @param {any} value - The value to set for the element's state.
+   * @returns {void}
+   */
+  public setElementState(elementId: string, key: string, value: any): void {
+    if (!this.elementStates.has(elementId)) {
+      this.elementStates.set(elementId, new Map<string, any>())
+    }
+
+    this.elementStates.get(elementId)?.set(key, value)
+  }
+
+  /**
+   * Remove the element state associated with a given element ID.
+   */
+  public deleteElementState(elementId: string): void {
+    this.elementStates.delete(elementId)
+  }
+
+  /**
+   * Remove the element state associated with a given element ID and the key.
+   */
+  public deleteElementStateForKey(elementId: string, key: string): void {
```

Comment:
I would probably merge this into the above method and make `key` optional. If `key` is not provided, delete the full state.